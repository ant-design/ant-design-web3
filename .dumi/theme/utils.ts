import { history, useSiteData } from 'dumi';
import type { ISidebarGroup } from 'dumi/dist/client/theme-api/types';

type ILocaleItem = ReturnType<typeof useSiteData>['locales'][0];

export function getTargetLocalePath({
  pathname = history.location.pathname,
  current,
  target,
}: {
  pathname?: string;
  current: ILocaleItem;
  target: ILocaleItem;
}) {
  const clearPath =
    'base' in current
      ? pathname.replace(current.base.replace(/\/$/, ''), '')
      : pathname.replace(new RegExp(`${current.suffix}$`), '');

  return 'base' in target
    ? `${target.base}${clearPath}`.replace(/^\/\//, '/')
    : `${clearPath}${target.suffix}`;
}

// 删除标题中的 HTML 标签
export const removeTitleCode = (title = '') => {
  return title.replace(/<\w+>.*<\/\w+>/g, '');
};

export const isExternalLinks = (link?: string): boolean => {
  return /^(\w+:)\/\/|^(mailto|tel):/.test(link ?? '');
};

export const handleFullSidebarData = (
  fullSidebarData: Record<string, ISidebarGroup[]>,
): Record<string, ISidebarGroup[]> => {
  const nextFullSidebarData: Record<string, ISidebarGroup[]> = {};
  Object.entries(fullSidebarData).forEach(([key, value]) => {
    const existedLevelOneNavList = Object.keys(nextFullSidebarData);
    if (key.split('/').length === 3) {
      // 二级目录
      const levelOneNavKey = `/${key.split('/')[1]}`;
      let _levelTwoNav = value[0]?.children?.find((item) => {
        const _frontmatterInfo = item.frontmatter?.nav;
        return typeof _frontmatterInfo === 'object' && typeof _frontmatterInfo?.second === 'object';
      });
      // 如果是 second 配置不为对象，直接获取其值即可
      if (!_levelTwoNav) {
        _levelTwoNav = value[0]?.children[0];
      }
      // 找找二级菜单名称以及顺序
      const _levelTwoNavInfo = _levelTwoNav?.frontmatter?.nav;
      const frontmatterNavSecond: { order: number; title?: string } = {
        order: 0,
        title: undefined,
      };
      if (typeof _levelTwoNavInfo !== 'string') {
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        frontmatterNavSecond.order = _levelTwoNavInfo?.second?.order ?? 0;
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        frontmatterNavSecond.title = _levelTwoNavInfo?.second?.title ?? _levelTwoNavInfo?.second;
      }

      if (existedLevelOneNavList.includes(levelOneNavKey)) {
        // 已存在，直接 push 为 children
        nextFullSidebarData[levelOneNavKey].push({
          title: frontmatterNavSecond.title,
          order: frontmatterNavSecond.order,
          children: value[0]?.children,
        });
      } else {
        nextFullSidebarData[levelOneNavKey] = value.map((item) => ({
          ...item,
          title: frontmatterNavSecond.title,
          order: frontmatterNavSecond.order,
        }));
      }
    } else if (existedLevelOneNavList.includes(key)) {
      nextFullSidebarData[key] = nextFullSidebarData[key].concat(value);
    } else {
      nextFullSidebarData[key] = value;
    }
  });
  return nextFullSidebarData;
};
