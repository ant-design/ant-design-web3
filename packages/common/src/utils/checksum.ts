import { isAddress, keccak256, toUtf8Bytes } from 'ethers';

import { ChainType } from '../types';

/**
 * @param address - An address string
 * @desc 将以太坊地址转换为经过校验和处理后的以太坊地址。
 * @returns	The checksum address
 * @example
 * ```ts
 * toChecksumAddressWithEVM('0x237dfeca906991dc570595a8176bfe65be3e8a7e');
 * > "0x237DFECA906991dc570595A8176bfE65Be3E8A7E"
 * ```
 */
export function toChecksumAddressWithEVM(address: string) {
  if (!isAddress(address)) {
    console.error('address is not a valid address', address);
    return address;
  }

  const lowerCaseAddress = address.toLowerCase().replace(/^0x/i, '');
  // 经过 isAddress 校验后 keccak256 不需要判断 EIP-1052
  const hash = keccak256(toUtf8Bytes(lowerCaseAddress));
  let ret = '';

  for (let i = 0; i < address.length; i++) {
    if (parseInt(hash[i], 16) >= 8) {
      ret += address[i].toUpperCase();
    } else {
      ret += address[i];
    }
  }
  return ret;
}

/**
 * @param address - 需要校验和的地址字符串
 * @param chain - 区块链类型
 * @desc 根据不同的区块链类型对地址进行校验和处理。目前支持 EVM 链的校验和地址生成。
 * @returns 经过校验和处理的地址，对于不支持的链类型返回原地址
 * @example
 * ```ts
 * checksumAddress('0x237dfeca906991dc570595a8176bfe65be3e8a7e', ChainType.EVM);
 * > "0x237DFECA906991dc570595A8176bfE65Be3E8A7E"
 *
 * checksumAddress('0x1234567890abcdef', ChainType.SOLANA);
 * > "0x1234567890abcdef" // 返回原地址，因为 Solana 不支持校验和
 * ```
 */
export const checksumAddress = (address: string, chain: ChainType) => {
  switch (chain) {
    case ChainType.EVM: {
      return toChecksumAddressWithEVM(address);
    }
    default:
      console.warn(`checksumAddress: unsupported chain type ${chain}`);
      return address;
  }
};
