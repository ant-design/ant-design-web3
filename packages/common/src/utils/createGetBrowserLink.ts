export const createGetBrowserLink = (url: string) => (address: string, type: string) => {
  if (type === 'address') {
    return `${url}/address/${address}`;
  }
  if (type === 'transaction') {
    return `${url}/tx/${address}`;
  }
  throw new Error(`getBrowserLink unsupported type ${type}`);
};
