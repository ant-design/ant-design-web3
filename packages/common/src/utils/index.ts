export * from './request';
export * from './format';
export * from './test-utils';
export * from './createGetBrowserLink';
export * from './warning';
export * from './checksum';
