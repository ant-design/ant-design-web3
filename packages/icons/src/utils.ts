// for fix https://github.com/ant-design/ant-design-web3/issues/1056
export function eraseDefaultProps(svgComponent: React.FC): React.FC {
  const defaultProps = svgComponent.defaultProps;
  return (props) => {
    const newProps = { ...defaultProps, ...props };
    return svgComponent(newProps);
  };
}
