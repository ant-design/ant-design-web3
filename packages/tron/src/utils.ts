/* v8 ignore start */
import { WalletReadyState } from '@tronweb3/tronwallet-abstract-adapter';

export const hasWalletReady = (readyState?: WalletReadyState) =>
  readyState === WalletReadyState.Found || readyState === WalletReadyState.Loading;
