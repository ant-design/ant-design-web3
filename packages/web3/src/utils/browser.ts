import copy from 'copy-to-clipboard';

export const getPlatform = () => {
  const userAgent = navigator.userAgent.toLowerCase();
  if (userAgent.includes('chrome')) {
    return 'Chrome';
  }
  if (userAgent.includes('firefox')) {
    return 'Firefox';
  }
  if (userAgent.includes('edge')) {
    return 'Edge';
  }
  if (userAgent.includes('safari')) {
    return 'Safari';
  }
  return 'Other';
};

export const writeCopyText = async (text: string): Promise<boolean> => {
  return copy(text);
};
