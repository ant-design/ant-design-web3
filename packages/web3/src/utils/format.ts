/**
 * @param address address
 * @param prefixProp addressPrefix from props
 * @param prefixContext addressPrefix from context
 */
export const fillWithPrefix = (
  address = '',
  prefixProp?: string | false,
  prefixContext?: string | false,
): string => {
  let prefix = undefined;

  // `0x` is the default prefix
  if (prefixProp === undefined && prefixContext === undefined) {
    prefix = '0x';
  }

  // has prefix be set
  else {
    // use props first
    if (prefixProp !== undefined) {
      prefix = prefixProp === false ? '' : prefixProp;
    }

    // use context
    else if (prefixContext !== undefined) {
      prefix = prefixContext === false ? '' : prefixContext;
    }
  }

  if (!prefix) return address;

  return address.startsWith(prefix) ? address : `${prefix}${address}`;
};

export const formatAddress = (address = '', groupSize = 4): string => {
  const formattedGroups = [];

  const has0x = address.startsWith('0x');
  /* v8 ignore next */
  const formatText = has0x ? address.slice(2) : address;

  for (let i = 0; i < formatText.length; i += groupSize) {
    const group = formatText.slice(i, i + groupSize);
    formattedGroups.push(group);
  }

  const formattedText = formattedGroups.join(' ');

  /* v8 ignore next */
  return has0x ? `0x ${formattedText}` : formattedText;
};

export const formatBalance = (value: bigint | number, decimals: number, fixed?: number): string => {
  const bigValue = typeof value === 'bigint' ? value : BigInt(value);
  const divisor = BigInt(10 ** decimals);
  const displayValue = bigValue / divisor;
  const fraction = bigValue % divisor;

  if (fraction === 0n && fixed === undefined) {
    return `${displayValue}`;
  }

  let fractionStr = fraction.toString().padStart(decimals, '0');
  if (fixed === undefined) {
    return `${displayValue}.${fractionStr.replace(/0+$/, '')}`;
  }
  fractionStr = fractionStr.substring(0, fixed).padEnd(fixed, '0');
  return `${displayValue}.${fractionStr}`;
};
