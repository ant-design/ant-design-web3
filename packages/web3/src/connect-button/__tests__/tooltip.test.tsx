import { fireEvent, render } from '@testing-library/react';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';

import { ConnectButton } from '..';
import { readCopyText } from '../../utils/test-utils';

describe('ConnectButton', () => {
  beforeEach(() => {
    vi.useFakeTimers();
  });
  afterEach(() => {
    vi.useRealTimers();
  });

  it('when tooltip is boolean, ant-tooltip not toBeNull', async () => {
    const App = () => {
      return (
        <ConnectButton
          account={{ address: '3ea2cfd153b8d8505097b81c87c11f5d05097c18' }}
          tooltip={true}
        />
      );
    };
    const { baseElement, rerender } = render(<App />);
    const btn = baseElement.querySelector('.ant-web3-address.ant-typography span')!;
    fireEvent.mouseEnter(btn);
    rerender(<App />);
    // When the tooltip's title is string, baseElement.outerHTML does not contain '.ant-tooltip'.
    // mouseEnterDelay defaults is 0.1s and waitFakeTimer is required.
    await vi.runAllTimersAsync();
    expect(baseElement.querySelector('.ant-tooltip')).not.toBeNull();
  });
  it('when mergedTitle does not exist, ant-tooltip toBeNull', () => {
    const { baseElement } = render(
      <ConnectButton
        account={{ address: '3ea2cfd153b8d8505097b81c87c11f5d05097c18' }}
        tooltip={{ open: true, copyable: true, title: undefined }}
      />,
    );
    expect(baseElement.querySelector('.ant-tooltip')).toBeNull();
  });

  it('display tooltip', () => {
    const { baseElement } = render(
      <ConnectButton
        account={{
          address: '3ea2cfd153b8d8505097b81c87c11f5d05097c18',
        }}
        tooltip={{ open: true }}
      />,
    );
    expect(baseElement.querySelector('.ant-tooltip')).not.toBeNull();
    expect(baseElement.querySelector('.ant-tooltip-inner')?.textContent?.trim()).toBe(
      '0x3ea2cfd153b8d8505097b81c87c11f5d05097c18',
    );
    expect(baseElement.querySelector('.anticon-copy')).not.toBeNull();
  });

  it('disabled copyable in tooltip', () => {
    const { baseElement } = render(
      <ConnectButton
        account={{ address: '3ea2cfd153b8d8505097b81c87c11f5d05097c18' }}
        tooltip={{ open: true, copyable: false }}
      />,
    );
    expect(baseElement.querySelector('.ant-tooltip')).not.toBeNull();
    expect(baseElement.querySelector('.ant-tooltip-inner')?.textContent?.trim()).toBe(
      '0x3ea2cfd153b8d8505097b81c87c11f5d05097c18',
    );
    expect(baseElement.querySelector('.anticon-copy')).toBeNull();
  });

  it('Enable copyable in tooltip', () => {
    const { baseElement } = render(
      <ConnectButton
        account={{ address: '3ea2cfd153b8d8505097b81c87c11f5d05097c18' }}
        tooltip={{ open: true, copyable: true }}
      />,
    );
    expect(baseElement.querySelector('.ant-tooltip')).not.toBeNull();
    expect(baseElement.querySelector('.ant-tooltip-inner')?.textContent?.trim()).toBe(
      '0x3ea2cfd153b8d8505097b81c87c11f5d05097c18',
    );
    expect(baseElement.querySelector('.anticon-copy')).toBeTruthy();
  });

  it('custom title in tooltip', () => {
    const { baseElement } = render(
      <ConnectButton
        account={{ address: '3ea2cfd153b8d8505097b81c87c11f5d05097c18' }}
        tooltip={{ open: true, title: 'aaaaaabbbbbbcccccc' }}
      />,
    );
    expect(baseElement.querySelector('.ant-tooltip')).not.toBeNull();
    expect(baseElement.querySelector('.ant-tooltip-inner')?.textContent?.trim()).toBe(
      'aaaaaabbbbbbcccccc',
    );
  });

  it('should not display tooltip when not custom title and without address in tooltip', () => {
    const { baseElement } = render(<ConnectButton tooltip />);
    expect(baseElement.querySelector('.ant-tooltip')).toBeNull();
  });

  it('should not display tooltip when without tooltip in tooltip', () => {
    const { baseElement } = render(<ConnectButton />);
    expect(baseElement.querySelector('.ant-tooltip')).toBeNull();
  });

  it('should copy text after click copy icon', async () => {
    const { baseElement } = render(
      <ConnectButton
        account={{ address: '3ea2cfd153b8d8505097b81c87c11f5d05097c18' }}
        tooltip={{ open: true }}
      />,
    );
    expect(baseElement.querySelector('.ant-tooltip')).not.toBeNull();
    expect(baseElement.querySelector('.ant-tooltip-inner')?.textContent?.trim()).toBe(
      '0x3ea2cfd153b8d8505097b81c87c11f5d05097c18',
    );
    expect(baseElement.querySelector('.anticon-copy')).not.toBeNull();
    fireEvent.click(baseElement.querySelector('.anticon-copy')!);
    await vi.waitFor(async () => {
      expect(baseElement.querySelector('.ant-message')).not.toBeNull();
      expect(baseElement.querySelector('.ant-message-notice-content')?.textContent).toBe('Copied!');
      await expect(readCopyText()).resolves.toBe('0x3ea2cfd153b8d8505097b81c87c11f5d05097c18');
    });
  });

  it('should copy text after click copy icon in custom title mode', async () => {
    const { baseElement } = render(
      <ConnectButton tooltip={{ open: true, title: 'aaaaaabbbbbbcccccc' }} />,
    );
    expect(baseElement.querySelector('.ant-tooltip')).not.toBeNull();
    expect(baseElement.querySelector('.ant-tooltip-inner')?.textContent?.trim()).toBe(
      'aaaaaabbbbbbcccccc',
    );
    expect(baseElement.querySelector('.anticon-copy')).not.toBeNull();
    fireEvent.click(baseElement.querySelector('.anticon-copy')!);
    await vi.waitFor(async () => {
      expect(baseElement.querySelector('.ant-message')).not.toBeNull();
      expect(baseElement.querySelector('.ant-message-notice-content')?.textContent?.trim()).toBe(
        'Copied!',
      );
      await expect(readCopyText()).resolves.toBe('aaaaaabbbbbbcccccc');
    });
  });
  it('should display formatted when pass format into tooltip', async () => {
    const { baseElement } = render(
      <ConnectButton
        account={{
          address: '0x3ea2cfd153b8d8505097b81c87c11f5d05097c18',
        }}
        tooltip={{ open: true, format: true }}
      />,
    );
    expect(baseElement.querySelector('.ant-tooltip')).not.toBeNull();
    expect(
      baseElement.querySelector('.ant-web3-connect-button-tooltip-content')?.textContent?.trim(),
    ).toBe('0x 3ea2 cfd1 53b8 d850 5097 b81c 87c1 1f5d 0509 7c18');
    expect(baseElement.querySelector('.anticon-copy')).not.toBeNull();
    fireEvent.click(baseElement.querySelector('.anticon-copy')!);
    await vi.waitFor(async () => {
      expect(baseElement.querySelector('.ant-message')).not.toBeNull();
      expect(baseElement.querySelector('.ant-message-notice-content')?.textContent?.trim()).toBe(
        'Copied!',
      );
      await expect(readCopyText()).resolves.toBe('0x3ea2cfd153b8d8505097b81c87c11f5d05097c18');
    });
  });
  it('should display formatted by custom formatter when pass format into tooltip', async () => {
    const { baseElement } = render(
      <ConnectButton
        account={{
          address: '0x3ea2cfd153b8d8505097b81c87c11f5d05097c18',
        }}
        tooltip={{ open: true, format: (address) => address.slice(0, 10) }}
      />,
    );
    expect(baseElement.querySelector('.ant-tooltip')).not.toBeNull();
    expect(
      baseElement.querySelector('.ant-web3-connect-button-tooltip-content')?.textContent?.trim(),
    ).toBe('0x3ea2cfd1');
    expect(baseElement).toMatchSnapshot();
  });
});
