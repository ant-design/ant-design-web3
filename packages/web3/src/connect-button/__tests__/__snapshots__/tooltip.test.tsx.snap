// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConnectButton > should display formatted by custom formatter when pass format into tooltip 1`] = `
<body>
  <div>
    <button
      class="ant-btn css-dev-only-do-not-override-ni1kz0 ant-btn-default ant-btn-color-default ant-btn-variant-outlined ant-web3-connect-button css-dev-only-do-not-override-ni1kz0 ant-tooltip-open"
      type="button"
    >
      <div
        class="ant-web3-connect-button-content"
      >
        <div
          class="ant-web3-connect-button-content-inner"
        >
          <div
            class="ant-web3-connect-button-text"
          >
            <span
              class="ant-typography ant-web3-address css-dev-only-do-not-override-ni1kz0 css-dev-only-do-not-override-ni1kz0"
            >
              <span>
                0x3ea2...7c18
              </span>
            </span>
          </div>
        </div>
      </div>
    </button>
  </div>
  <div>
    <div
      class="ant-tooltip ant-zoom-big-fast-appear ant-zoom-big-fast-appear-prepare ant-zoom-big-fast ant-web3-connect-button-tooltip css-dev-only-do-not-override-ni1kz0 css-dev-only-do-not-override-ni1kz0 ant-tooltip-placement-top"
      style="--arrow-x: 0px; --arrow-y: 0px; left: -1000vw; top: -1000vh; box-sizing: border-box;"
    >
      <div
        class="ant-tooltip-arrow"
        style="position: absolute; bottom: 0px; left: 0px;"
      />
      <div
        class="ant-tooltip-content"
      >
        <div
          class="ant-tooltip-inner"
          role="tooltip"
        >
          <div
            class="ant-web3-connect-button-tooltip-title"
          >
            Wallet address
             
            <span
              aria-label="copy"
              class="anticon anticon-copy"
              role="img"
              tabindex="-1"
              title="Copy Address"
            >
              <svg
                aria-hidden="true"
                data-icon="copy"
                fill="currentColor"
                focusable="false"
                height="1em"
                viewBox="64 64 896 896"
                width="1em"
              >
                <path
                  d="M832 64H296c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h496v688c0 4.4 3.6 8 8 8h56c4.4 0 8-3.6 8-8V96c0-17.7-14.3-32-32-32zM704 192H192c-17.7 0-32 14.3-32 32v530.7c0 8.5 3.4 16.6 9.4 22.6l173.3 173.3c2.2 2.2 4.7 4 7.4 5.5v1.9h4.2c3.5 1.3 7.2 2 11 2H704c17.7 0 32-14.3 32-32V224c0-17.7-14.3-32-32-32zM350 856.2L263.9 770H350v86.2zM664 888H414V746c0-22.1-17.9-40-40-40H232V264h432v624z"
                />
              </svg>
            </span>
          </div>
          <div
            class="ant-web3-connect-button-tooltip-content"
          >
            0x3ea2cfd1
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
